aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,560 @@
+
+![Admin MFA authentication](mfa/_static/identitystandalonemfa_01.png)
+
+When the user clicks the admin link, the user is redirected to the MFA enable view.
+
+![Admin activate MFA authentication](mfa/_static/identitystandalonemfa_02.png)
+
+## Send MFA signin requirement to OpenID Connect server 
+
+The `acr_values` parameter can used to pass the mfa required value from the client to the server in an authentication request. 
+
+> [!NOTE]
+> The **acr_values** parameter needs to be handled on the Open ID Connect server for this to work.
+
+### OpenID Connect ASP.NET Core client
+
+The ASP.NET Core Razor Pages Open ID Connnect client app uses the `AddOpenIdConnect` method to login to the Open ID Connect server. The `acr_values` parameter is set with the `mfa` value and sent with the authentication request. The `OpenIdConnectEvents` is used to add this.
+
+See [Authentication Method Reference Values](https://tools.ietf.org/html/draft-ietf-oauth-amr-values-08) for recommended acr_values values.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+	services.AddAuthentication(options =>
+	{
+		options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+		options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+	})
+	.AddCookie()
+	.AddOpenIdConnect(options =>
+	{
+		options.SignInScheme = "Cookies";
+		options.Authority = "<OpenID Connect server URL>";
+		options.RequireHttpsMetadata = true;
+		options.ClientId = "<OpenID Connect client ID>";
+		options.ClientSecret = "<>";
+		options.ResponseType = "code id_token"; // Code with PKCE can also be used here
+		options.Scope.Add("profile");
+		options.Scope.Add("offline_access");
+		options.SaveTokens = true;
+		options.Events = new OpenIdConnectEvents
+		{
+			OnRedirectToIdentityProvider = context =>
+			{
+				context.ProtocolMessage.SetParameter("acr_values", "mfa");
+
+				return Task.FromResult(0);
+			}
+		};
+	});
+
+```
+
+### Example OpenID Connect IdentityServer 4 server with ASP.NET Core Identity
+
+On the OpenID Connect server, which is implemented using ASP.NET Core Identity with MVC views, a new view *ErrorEnable2FA.cshtml* is created and added.
+
+This view will be displayed if the Identity comes from an application which requires MFA but the user has not activated this in Identity. The view informs the user, and adds a link to activate this.
+
+```cshtml
+@{
+    ViewData["Title"] = "ErrorEnable2FA";
+}
+
+<h1>The client application requires you to have MFA enabled. Enable this, try login again.</h1>
+
+<br />
+
+You can enable MFA to login here:
+
+<br />
+
+<a asp-controller="Manage" asp-action="TwoFactorAuthentication">Enable MFA</a>
+```
+
+In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` is accessed using the AcrValues. As the client sent this with **mfa** set, this can then be checked.
```

Comment:
```suggestion
In the `Login` method, the `IIdentityServerInteractionService` interface implementation _interaction is used to access the Open ID Connnect request parameters. The `acr_values` parameter is accessed using the `AcrValues` property. As the client sent this with `mfa` set, this can then be checked.
```